File: apps/chaoxing_app/lib/services/reminder_rules.dart
import '../models/sync_item.dart';

enum ReminderIntensity { low, high }

class ReminderHistory {
  const ReminderHistory(this.sent);

  const ReminderHistory.empty() : sent = const {};

  final Map<String, DateTime> sent;

  bool contains(String key) => sent.containsKey(key);

  ReminderHistory markSent(String key, DateTime sentAt) {
    return ReminderHistory({...sent, key: sentAt});
  }

  ReminderHistory prune(
    DateTime now, {
    Duration retention = const Duration(days: 90),
    int maximumEntries = 1000,
  }) {
    final oldest = now.subtract(retention);
    final newest = now.add(const Duration(days: 1));
    final entries =
        sent.entries
            .where(
              (entry) =>
                  !entry.value.isBefore(oldest) && !entry.value.isAfter(newest),
            )
            .toList()
          ..sort((left, right) {
            final timeOrder = right.value.compareTo(left.value);
            return timeOrder != 0 ? timeOrder : left.key.compareTo(right.key);
          });
    return ReminderHistory(
      Map.fromEntries(
        entries.take(maximumEntries.clamp(0, entries.length).toInt()),
      ),
    );
  }

  factory ReminderHistory.fromJson(Map<String, dynamic> json) {
    final raw = json['sent'];
    if (raw is! Map) {
      return const ReminderHistory.empty();
    }
    return ReminderHistory(
      raw.map(
        (key, value) => MapEntry(
          key.toString(),
          value is String
              ? DateTime.tryParse(value)?.toLocal() ?? DateTime(1970)
              : DateTime(1970),
        ),
      ),
    );
  }

  Map<String, dynamic> toJson() => {
    'sent': sent.map((key, value) => MapEntry(key, value.toIso8601String())),
  };
}

class PlannedReminder {
  const PlannedReminder({
    required this.key,
    required this.item,
    required this.ruleId,
    required this.triggerAt,
    required this.intensity,
  });

  final String key;
  final SyncItem item;
  final String ruleId;
  final DateTime triggerAt;
  final ReminderIntensity intensity;
}

const _rules = [
  (
    id: 'due-24h',
    beforeDue: Duration(hours: 24),
    intensity: ReminderIntensity.low,
  ),
  (
    id: 'due-2h',
    beforeDue: Duration(hours: 2),
    intensity: ReminderIntensity.high,
  ),
];

List<PlannedReminder> planReminders({
  required List<SyncItem> items,
  required ReminderHistory history,
  required DateTime now,
}) {
  final plans = <PlannedReminder>[];
  for (final item in items) {
    final dueAt = item.dueAt;
    if (dueAt == null || !dueAt.isAfter(now)) {
      continue;
    }
    for (final rule in _rules) {
      final key = [
        item.id,
        item.kind.name,
        rule.id,
        dueAt.toIso8601String(),
      ].join('|');
      if (!history.contains(key)) {
        plans.add(
          PlannedReminder(
            key: key,
            item: item,
            ruleId: rule.id,
            triggerAt: dueAt.subtract(rule.beforeDue),
            intensity: rule.intensity,
          ),
        );
      }
    }
  }
  plans.sort((left, right) {
    final timeOrder = left.triggerAt.compareTo(right.triggerAt);
    return timeOrder != 0 ? timeOrder : left.key.compareTo(right.key);
  });
  return plans;
}

File: packages/chaoxing-domain/src/reminder-rules.ts
import { type SyncItem } from "./sync-item";
import { addDays, addHours, toLocalIso8601 } from "./time";

export const ReminderIntensity = {
  low: "low",
  high: "high",
} as const;

export type ReminderIntensity =
  (typeof ReminderIntensity)[keyof typeof ReminderIntensity];

export type ReminderHistory = {
  sent: Readonly<Record<string, Date>>;
};

export type PlannedReminder = {
  key: string;
  item: SyncItem;
  ruleId: string;
  triggerAt: Date;
  intensity: ReminderIntensity;
};

const rules = [
  {
    id: "due-24h",
    beforeDueHours: 24,
    intensity: ReminderIntensity.low,
  },
  {
    id: "due-2h",
    beforeDueHours: 2,
    intensity: ReminderIntensity.high,
  },
] as const;

export function emptyReminderHistory(): ReminderHistory {
  return { sent: {} };
}

export function reminderHistoryContains(
  history: ReminderHistory,
  key: string,
): boolean {
  return Object.prototype.hasOwnProperty.call(history.sent, key);
}

export function markReminderSent(
  history: ReminderHistory,
  key: string,
  sentAt: Date,
): ReminderHistory {
  return { sent: { ...history.sent, [key]: sentAt } };
}

export function pruneReminderHistory(
  history: ReminderHistory,
  now: Date,
  options: { retentionDays?: number; maximumEntries?: number } = {},
): ReminderHistory {
  const retentionDays = options.retentionDays ?? 90;
  const maximumEntries = options.maximumEntries ?? 1000;
  const oldest = addDays(now, -retentionDays);
  const newest = addDays(now, 1);
  const entries = Object.entries(history.sent)
    .filter(([, sentAt]) => sentAt >= oldest && sentAt <= newest)
    .sort(([leftKey, leftAt], [rightKey, rightAt]) => {
      const timeOrder = rightAt.getTime() - leftAt.getTime();
      return timeOrder !== 0 ? timeOrder : leftKey.localeCompare(rightKey);
    });
  const limit = Math.min(Math.max(maximumEntries, 0), entries.length);
  return {
    sent: Object.fromEntries(entries.slice(0, limit)),
  };
}

export function reminderKey(item: SyncItem, ruleId: string, dueAt: Date): string {
  return [item.id, item.kind, ruleId, toLocalIso8601(dueAt)].join("|");
}

export function planReminders(input: {
  items: readonly SyncItem[];
  history: ReminderHistory;
  now: Date;
}): PlannedReminder[] {
  const plans: PlannedReminder[] = [];
  for (const item of input.items) {
    const dueAt = item.dueAt;
    if (dueAt === null || dueAt.getTime() <= input.now.getTime()) {
      continue;
    }
    for (const rule of rules) {
      const key = reminderKey(item, rule.id, dueAt);
      if (!reminderHistoryContains(input.history, key)) {
        plans.push({
          key,
          item,
          ruleId: rule.id,
          triggerAt: addHours(dueAt, -rule.beforeDueHours),
          intensity: rule.intensity,
        });
      }
    }
  }
  plans.sort((left, right) => {
    const timeOrder = left.triggerAt.getTime() - right.triggerAt.getTime();
    return timeOrder !== 0 ? timeOrder : left.key.localeCompare(right.key);
  });
  return plans;
}

export function collectDueReminders(input: {
  items: readonly SyncItem[];
  history: ReminderHistory;
  now: Date;
}): PlannedReminder[] {
  return planReminders(input).filter(
    (plan) => plan.triggerAt.getTime() <= input.now.getTime(),
  );
}
